service/common/src/main/java/org/apache/polaris/service/auth/DefaultAuthenticator.java

```diff
@@ -88,17 +92,19 @@ protected Optional<AuthenticatedPolarisPrincipal> getPrincipal(DecodedToken toke
           .log("Unable to authenticate user with token");
       throw new ServiceFailureException("Unable to fetch principal entity");
     }
-    if (principal == null) {
-      LOGGER.warn(
-          "Failed to resolve principal from tokenInfo client_id={}", tokenInfo.getClientId());
+    if (principal == null || principal.getType() != PolarisEntityType.PRINCIPAL) {
+      LOGGER.warn("Failed to resolve principal from credentials={}", credentials);
       throw new NotAuthorizedException("Unable to authenticate");
     }
 
+    LOGGER.debug("Resolved principal: {}", principal);
+
+    boolean allRoles = credentials.getPrincipalRoles().contains(PRINCIPAL_ROLE_ALL);
+
     Set<String> activatedPrincipalRoles = new HashSet<>();
-    // TODO: Consolidate the divergent "scopes" logic between test-bearer-token and token-exchange.
-    if (tokenInfo.getScope() != null && !tokenInfo.getScope().equals(PRINCIPAL_ROLE_ALL)) {
+    if (!allRoles) {
       activatedPrincipalRoles.addAll(
-          Arrays.stream(tokenInfo.getScope().split(" "))
+          credentials.getPrincipalRoles().stream()
               .map(
                   s -> // strip the principal_role prefix, if present
                   s.startsWith(PRINCIPAL_ROLE_PREFIX)
```

Comment:
I believe this is dangerous and should be changed. For example, if the roles were initially:
`["PRINCIPAL_ROLE:catalog-admin', "service-admin"]`, this would be transformed into: `["catalog-admin', "service-admin"]`, potentially requesting the `service-admin` role, but that likely wasn't intentional.

I think it would be safer to discard all roles that don't have the `PRINCIPAL_ROLE:` prefix.

Reply:
+1 to filter

Reply:
BTW I also think that the responsibility of stripping off the prefix should be transferred from this class to `DefaultRolesMapper`. But we can do this in a follow-up PR.